flight/PiOS/Common/pios_mpu.c

```diff
@@ -0,0 +1,1118 @@
+
+#ifdef PIOS_INCLUDE_MPU_MAG
+/**
+ * @brief Allocate a a magnetometer queue
+ */
+static bool PIOS_MPU_Mag_Alloc(struct pios_mpu_dev *mpu_dev);
+static int32_t PIOS_MPU_Mag_Probe(void);
+#endif // PIOS_INCLUDE_MPU_MAG
+/**
+ * @brief Validate the handle to the device
+ * @returns 0 for valid device or -1 otherwise
+ */
+static int32_t PIOS_MPU_Validate(struct pios_mpu_dev *dev);
+/**
+ * @brief Initialize the MPU gyro & accel registers
+ * @param[in] pios_mpu_cfg struct to be used to configure sensor.
+ * @return 0 if successful
+ */
+static int32_t PIOS_MPU_Config(struct pios_mpu_cfg const *cfg);
+static void PIOS_MPU_Task(void *parameters);
+static int32_t PIOS_MPU_ReadReg(uint8_t reg);
+static int32_t PIOS_MPU_WriteReg(uint8_t reg, uint8_t data);
+
+#if defined(PIOS_INCLUDE_SPI) || defined(__DOXYGEN__)
+/**
+ * @brief Claim the SPI bus for the communications and select this chip
+ * \param[in] flag controls if low speed access for control registers should be used
+ * @return 0 if successful, -1 for invalid device, -2 if unable to claim bus
+ */
+static int32_t PIOS_MPU_ClaimBus(bool lowspeed);
+/**
+ * @brief Release the SPI bus for the communications and end the transaction
+ * \param[in] must be true when bus was claimed in lowspeed mode
+ * @return 0 if successful
+ */
+static int32_t PIOS_MPU_ReleaseBus(bool lowspeed);
+/**
+ * @brief Probe the SPI bus for an MPU device
+ * @param[out] Detected device type, only valid on success
+ * @retval 0 success, -1 failure
+ */
+static int32_t PIOS_MPU_SPI_Probe(enum pios_mpu_type *detected_device);
+#endif // defined(PIOS_INCLUDE_SPI) || defined(__DOXYGEN__)
+
+#if defined(PIOS_INCLUDE_I2C) || defined(__DOXYGEN__)
+/**
+ * @brief Probe the I2C bus for an MPU device
+ * @param[out] Detected device type, only valid on success
+ * @retval 0 success, -1 failure
+ */
+static int32_t PIOS_MPU_I2C_Probe(enum pios_mpu_type *detected_device);
+#endif // defined(PIOS_INCLUDE_I2C) || defined(__DOXYGEN__)
+
+
+static struct pios_mpu_dev *PIOS_MPU_Alloc(const struct pios_mpu_cfg *cfg)
+{
+	struct pios_mpu_dev *dev;
+
+	dev = (struct pios_mpu_dev *)PIOS_malloc(sizeof(*mpu_dev));
+	if (!dev)
+		return NULL;
+
+	dev->magic = PIOS_MPU_DEV_MAGIC;
+
+	dev->accel_queue = PIOS_Queue_Create(PIOS_MPU_QUEUE_LEN, sizeof(struct pios_sensor_accel_data));
+	if (dev->accel_queue == NULL) {
+		PIOS_free(dev);
+		return NULL;
+	}
+
+	dev->gyro_queue = PIOS_Queue_Create(PIOS_MPU_QUEUE_LEN, sizeof(struct pios_sensor_gyro_data));
+	if (dev->gyro_queue == NULL) {
+		PIOS_Queue_Delete(dev->accel_queue);
+		PIOS_free(dev);
+		return NULL;
+	}
+
+	dev->data_ready_sema = PIOS_Semaphore_Create();
+	if (dev->data_ready_sema == NULL) {
+		PIOS_Queue_Delete(dev->accel_queue);
+		PIOS_Queue_Delete(dev->gyro_queue);
+		PIOS_free(dev);
+		return NULL;
+	}
+
+	return dev;
+}
+
+static int32_t PIOS_MPU_Validate(struct pios_mpu_dev *dev)
+{
+	if (dev == NULL)
+		return -1;
+	if (dev->magic != PIOS_MPU_DEV_MAGIC)
```

Comment:
i hate these conventions.  we need to end them.  but not for this PR

(trying to tolerate / error-out sanely when.. someone is passing us something that's not the IMU as the IMU?)